### ISSUE TITLE
Handle INFO messages received from the cluster

### ISSUE THREAD

**Issue by mmmries** [mmmries/gnat#54]
@quixoten pointed out [in a previous PR](https://github.com/mmmries/gnat/pull/53#issuecomment-299318419) that when the cluster topology changes (nodes enter or leave?) nats sends `INFO` messages with additional information about where each node lives. This way you could do a rolling set of restarts and give the new nodes totally different addresses while keeping all clients online.

We should add support for these messages and somehow update our list of available nodes.

**Comment by mmmries** [comment]
@quixoten do you know if the username/password/tls etc settings are always required to be the same across a cluster? Seems like that's what most people will do, but just wondering if it is required

**Comment by quixoten** [comment]
It is not required. I was able to cluster two gnatsd servers with different TLS settings. One enabled and one disabled. I think the same is likely true of the auth settings because the cluster config has separate auth settings, which could be an interesting way of bridging clients in different security zones.

Not sure what this looks like to implement. Maybe some layer that will toss out invalid server uris and than crash altogether when it runs out? Could a layer above that then collect a fresh list of server uris, e.g., from etcd, and then restart the next layer down?

**Comment by mmmries** [comment]
Hmmmm, I wonder if we should make the client code handle this problem? Maybe we let the client pass in a function that is used to select a host each time we try to establish a connection?

For a basic fixed-set of nodes that function could just sample from the list each time, but for someone using `etcd` they could call down into that and then generate the connection settings?

Do we get notifications from the brokers when a node leaves the cluster?

**Comment by mmmries** [comment]
I tried to find some more details about this and from what I can see the `INFO` message for nodes joining the cluster only applies to the [server protocol](http://nats.io/documentation/internals/nats-server-protocol/) which is used between the brokers. But the [client protocol](http://nats.io/documentation/internals/nats-protocol/) says that `INFO` messages are only presented when you open a connection.

> As soon as the server accepts a connection from the client, it will send information about itself and the configuration and security requirements that are necessary for the client to successfully authenticate with the server and exchange messages.

**Comment by quixoten** [comment]
I agree it's not documented very well, maybe because it's a newer feature, but it is part of the client protocol; however, the client must declare support for it in its CONNECT message. See the documentation for `CONNECT`.

**Comment by quixoten** [comment]
Here's where some other clients handle this:

[pure-ruby-nats](https://github.com/nats-io/pure-ruby-nats/blob/4a60b9fe1fbfcd7bc4b62feb4d9aca7e728b9e00/lib/nats/io/client.rb#L486)
[go-nats](https://github.com/nats-io/go-nats/blob/85cb33631e90220ed9d2e23dc855f65a2efc9737/nats.go#L1749-L1750)
[cnats](https://github.com/nats-io/cnats/blob/6f7cc22fb691591597dcf6ce7febf0b09c3f8ee1/src/conn.c#L502-L506)
[java-nats](https://github.com/nats-io/java-nats/blob/89aac652a8279c92dcaa91042ccc036ae147c2dd/src/main/java/io/nats/client/ConnectionImpl.java#L769)

Here is the server side implementation:

[update info and broadcast to clients](https://github.com/nats-io/gnatsd/blob/55404c70b0418a7017d05c504cf4a647afe91e25/server/route.go#L160-L162)
